src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -861,6 +861,84 @@ void StorageReplicatedMergeTree::createReplica(const StorageMetadataPtr & metada
 
     LOG_DEBUG(log, "Creating replica {}", replica_path);
 
+    const String local_metadata = ReplicatedMergeTreeTableMetadata(*this, metadata_snapshot).toString();
+    const String local_columns = metadata_snapshot->getColumns().toString();
+    const String local_metadata_version = toString(metadata_snapshot->getMetadataVersion());
+
+    /// It is possible for the replica to fail after creating ZK nodes without saving local metadata.
+    /// Because of that we need to check whether the replica exists and is newly created.
+    /// For this we check that all nodes exist, the metadata of the table is the same, and other nodes are not modified.
+
+    std::vector<String> paths_exists = {
+        replica_path,
+        replica_path + "/host",
+        replica_path + "/log_pointer",
+        replica_path + "/queue",
+        replica_path + "/parts",
+        replica_path + "/flags",
+        replica_path + "/is_lost",
+        replica_path + "/metadata",
+        replica_path + "/columns",
+        replica_path + "/metadata_version",
+        replica_path + "/mutation_pointer",
+        replica_path + "/min_unprocessed_insert_time",
+        replica_path + "/max_processed_insert_time",
+        replica_path + "/mutation_pointer"
+    };
+
+    auto response_exists = zookeeper->tryGet(paths_exists);
+    size_t response_num = 0;
+
+    if (response_exists[response_num++].error == Coordination::Error::ZOK)
+    {
+        bool all_nodes_exist = true;
+
+        for (size_t i = 0; i < response_exists.size(); ++i)
+        {
+            if (response_exists[i].error != Coordination::Error::ZOK)
+            {
+                all_nodes_exist = false;
+                break;
+            }
+        }
+
+        if (all_nodes_exist)
+        {
+            const auto & zk_host                        = response_exists[response_num++].data;
+            const auto & zk_log_pointer                 = response_exists[response_num++].data;
+            const auto & zk_queue                       = response_exists[response_num++].data;
+            const auto & zk_parts                       = response_exists[response_num++].data;
+            const auto & zk_flags                       = response_exists[response_num++].data;
+            const auto & zk_is_lost                     = response_exists[response_num++].data;
```

Comment:
Should we still call `zookeeper->tryGet(zookeeper_path + "/replicas", replicas_value, &replicas_stat)` in the beginning of this method? Because I saw this path we assume it already there before this method executed. And the `is_lost_value` is determined based on `is_lost_value = replicas_stat.numChildren ? "1" : "0"`. Here we relaxed the condition, either 0 or 1 is okay.

Reply:
> Should we still call zookeeper->tryGet(zookeeper_path + "/replicas", replicas_value, &replicas_stat) in the beginning of this method? Because I saw this path we assume it already there before this method executed.

Yep, and I also don't see the reason to check it here since if it does not exist all the paths that we get will return an error `Coordination::Error::ZNONODE` the same way as if it existed but a replica didn't.

`is_lost_value` can be `"0"` if it is the first replica and `"1"` if it's not. We want to cover both cases.